### DISCUSSION TITLE
getting_started/step_by_step/scripting_first_script

### DISCUSSION THREAD

**Discussion by giscus** [godotengine/godot-docs-user-notes#52] [User-contributed notes]
# getting_started/step_by_step/scripting_first_script

In this lesson, you will code your first script to make the Godot icon turn in circles using GDScript. As we mentioned in the introduction, we assume you have programming foundations. The equivalen...

https://docs.godotengine.org/en/stable/getting_started/step_by_step/scripting_first_script.html

**Comment by mrwensveen** [comment]
It works but my animation was choppy at times. Disabling VSync worked for me, see: https://docs.godotengine.org/en/stable/tutorials/rendering/jitter_stutter.html

**Comment by CyrilDemand** [comment]
Hi, for the Hello World part in C#, it didn't work for me until I call GD.Print in a _Ready method.
public override void _Ready()
{
	GD.Print("Hello, world!");
}

**Reply by mokgul** [reply]
It works with _Ready() because the engine calls that method when the class instance is creating, Another way is to place the code in the class constructor (or call a method that executes that code). Constructors are called when an instance is created and it will work.

```cs
public partial class Sprite2d : Sprite2D
{
    public Sprite2d()
    {
        GD.Print("Hello, world!");
    }
}
```

or

```cs
public partial class Sprite2d : Sprite2D
{
    public Sprite2d()
    {
        PrintText()
    }

    public override void PrintText()
    {
        GD.Print("Hello, world!");
    ]
```

**Comment by InfieldTriple** [comment]
I personally found that sprite2d.gd script to be a bit unintuitive and it could use some changes. It might be due to my physics/math background. I've written a separate version of this script to separate the rotation (the image turning upside down) from the circular motion (motion of the centre of the axis around some central point).

```gdscript
extends Sprite2D

# circular motion angular speed
var angular_speed = PI/2

# rotation angular speed
var rotate_speed = 2*PI

# radius of rotation
var radius = 100

# Time vector
var mytime = 0


func _process(delta):

    # Time marches on
	mytime += delta

    # x (left/right) and y (up/down) speed
	var xvel = -radius * sin(2*PI*mytime/angular_speed)
	var yvel = radius * cos(2*PI*mytime/angular_speed)

    # Velocity vector
	var vel = xvel * Vector2.RIGHT + yvel * Vector2.UP 
	
	position += vel * delta
	rotation += rotate_speed * delta
```

This new version of the function allows you to set the motion dimensions more directly. The radius of the circular motion and the rotation and circular motion speeds separately.

I suspect there is a better way to do this in godot. I simply evaluated the derivative of the position vector of an object in circular motion my hand and hard coded it.

Something I think is missing from this doc (not sure if this qualifies for an issue to be raised) is an explanation of how sprite2d.gd works to rotate the sprite. Something like:

> The sprite is forced in the direction `Vector2.UP` at speed `speed`. The position is changed by multiplying `velocity` by the time step `delta`, creating a new position vector and adding it to the previous position. The sprite follows a circular path by rotating the `Vector2.UP` vector using the `Vector2.UP.rotated` function where it supplies the variable `rotation` as an argument so that the vector `Vector2.UP` is rotated the exact same amount as the sprite. Then the sprite is given a velocity in that direction. Since at every time step this is adjusted, the sprite is moved forward slightly at this new angle resulting in circular motion.

I tried to decouple the rotation and circulation motion as follows by defining two angular_speed variables. This worked however, the radius still changes with angular_speed2 which makes sense physically but can be unintuitive. Thankfully, that only matters for anyone reading the code and not the player.

```gdscript
extends Sprite2D

var speed = 400
var angular_speed1 = PI
var angular_speed2 = 2*PI
var myrotation = 0

func _init():
	myrotation = rotation

func _process(delta):
	rotation += angular_speed1 * delta
	myrotation += angular_speed2 * delta
	var velocity = Vector2.UP.rotated(myrotation) * speed

	position += velocity * delta
```

**Reply by jsonjunkie** [reply]
i think you are over-thinking and over-complicating the sprite script. while it could be useful to have more control over how the icon moves and rotates, the extra bits of your scripts feel out of scope for this lesson

when you consider that every tick increases the rotation of the icon and that same rotation is applied to the velocity vector, the existing example might feel more intuitive

**Reply by scotthawkins** [reply]
I find what people see as intuitive is very interesting.  I really appreciate InfieldTriple's detailed explication of how the example works.  But I suspect their version of doing the manual trig implementation might not be intuitive for most people.  On the other hand, while the example's generation and scaling of the unit vector is probably fairly intuitive for someone with a graphical programming background, it might be a little confounding for a true beginner who is new to all of this.  Since this is an introduction for a beautifully accessible tool for all levels, a bit of simple explanation highlighting the use of the unit UP vector, rotation, and scaling would probably serve a valuable portion of the users of this tutorial.  I hope that people will feel invited to pull the example apart and work to make themselves understand how the parts work.

**Comment by pablomederos** [comment]
If you are new to C# programming, the **Hello World** code example
```csharp
public MySprite2D()
{
    GD.Print("Hello, world!");
}
```

is placed in a constructor. Don't put it in the class body 'cause it won't work.

**Reply by ska00** [reply]
If you want the sprite to rotate in a circle with a given radius you'll need to adjust the angular speed according to the equation: $\omega = v/r$ where $\omega$ is the angular speed.
(Make sure the speed or radius have a decimal point so that the division doesn't truncate the result)

```gdscript
var radius = 100
var speed = 400.0
# var angular_speed = PI
var angular_speed = speed / radius 
```

The sprite will still rotate with the same speed, so a larger radius means it will take longer to complete 1 rotation. If you want to keep the time it takes to complete 1 rotation the same do this:

```gdscript
var radius = 100
var speed = 400.0 * (radius / 100.0)
var angular_speed = speed / radius
```

That way the speed increases for larger radii and decreases for smaller radii keeping the period constant. Note that by keeping the radius fixed, changing the linear speed will indirectly change the angular speed and vice versa.

**Reply by ska00** [reply]
I meant to reply to the comment above...

**Reply by BytesBlade** [reply]
I did the same but there was no output in the panel below. I see this error in the Inspector box: "This inspector might be out of date. Please build the C# project".

**Comment by DarthRevan39** [comment]
Hi, I had an issue with the icon displaying.

I was just running through this on Godot 4.3 and my Sprite2D was not set to be visible by default. Under the Inspector tab I had to go down to Visibility under CanvasItem and toggle visibility to on before I could see the Godot icon in the scene.

**Reply by Calinou** [reply]
If the sprite wasn't visible, it likely means you clicked on the eye icon besides it in the Scene tree dock at some point. **Visible** is `true` by default.

**Reply by DarthRevan39** [reply]
That is very possible! I didn't know that eye icon was there, but now I see it!

Thank you!

**Comment by TheEagerCoder** [comment]
I did this and then started messing around and experimenting. One thing I did was ramp up speed to 2000. After doing this, the sprite started to have a blur behind it. Not sure if this is an intended feature, a biological issue (my eyes make it look like a blur because of its speed), or if perhaps a bug?

**Reply by Calinou** [reply]
The blur is likely caused by display ghosting, which is a hardware issue. Higher movement speeds will make the issue more obvious. Check if the ghosting is also visible on [TestUFO](https://www.testufo.com/).

Some ways to mitigate ghosting:

- Ensure the game is rendering at the same framerate as the display's refresh rate (or higher).
- Use a brighter background, as dark backgrounds are prone to smearing on VA panels.
- Check your monitor's OSD settings and enable overdrive. You don't want to set it too high though, as excessive overdrive can lead to reverse ghosting where bright trails appear behind moving objects.
  - On most displays, the second highest setting is usually the right one to pick when using the display at its maximum refresh rate.
- Check your monitor's OSD settings and enable [motion blur reduction](https://blurbusters.com/faq/motion-blur-reduction/). This is the most effective approach, but it's not available on all monitors and comes with significant limitations (such as not being usable with VRR on most displays).

**Comment by jakeometer** [comment]
Having issues, the sprite doesn't spin, it orbits the top left hand corner, entering frame top right exiting bottom left. This is with AND without the "velocity" and "position" lines of code. Have I missed something?

**Reply by ManBeaRTo** [reply]
I have the same issue

**Reply by GreNinjaFour0Four** [reply]
I fixed mine by putting Camera 2d node and dragging the camera to center

**Reply by SriramAnanth2912** [reply]
I used this code to center the sprite2D w.r.t viewport:
```gdscript
var screen_size: Vector2 ## screen size of the viewport

func _ready() -> void:
	screen_size = get_viewport_rect().size
	global_position = screen_size / 2
```

**Reply by smartcode444** [reply]
I think you should move the godot sprite to the center of the delimited rectangle on your 2d workspace. By default, the sprite is positioned at the top-left corner, which is what is causing your problem.

**Reply by VadimKovalenko** [reply]
The sprite was orbiting because the image was added as a separate child node far from the parent’s origin instead of being assigned as the parent's (Sprite2D) actual texture. To fix it, you must drag the image from the FileSystem directly into the Texture slot of the Sprite2D node in the Inspector, then manually position that node in the 2D viewport.

**Comment by madstork91** [comment]
Literally the first part of the code
```gdscript
extends Sprite2D

func _init():
	print("hello, world!")
```

Returns errors.
From output:
**Another resource is loaded from path 'res://sprite_2d.gd' (possible cyclic resource inclusion).**

from the editor:
**Cannot run the script because it doesn't extend EditorScript.**
**Cannot run the script because it's not a tool script (add the @tool annotation at the top).**

When following the editor errors, it give another error saying you can only extend one... so I remove extends Sprite2D...

Am I doing something wrong? Or is the above wrong?

**Reply by nameless-creator** [reply]
I had the same thing, the tutorial says to "run it" which I had interpreted as "run the script" rather than what it actually means, which is "play the scene".

**Comment by Hiina4872** [comment]
I followed the directions exactly but the icon will not move

**Reply by Binasaurus** [reply]
1. Try saving first, I guess it should work
2. Just check if none of the other variables are in caps (only Vector2.UP) is in caps
3. In other languages we usually use semicolons(;) which is not the case here, took time adjusting for me too
Hope this helps

**Comment by copilat1** [comment]
(4.4) so, I tried to type the last 2 lines of code and the last line was an error. so I tried copying and pasting the last 2 lines and there was still an error. so I tried copying and pasting the entire script and the error was gone. so I went to the old script and modified it to be the same as the pasted and there was an error on the second to last line. IDK what this is, but it works.

**Reply by BananaPi515** [reply]
Had the same problem

**Reply by SriramAnanth2912** [reply]
May be indentation issue?

**Comment by antonprv** [comment]
Tried to adapt the code slightly to use angles and convert them to radians after, but icon just flies off the screen. What am I doing wrong here?

```csharp
using Godot;

namespace Prototype1.Scripts;

public partial class MySprite2D : Sprite2D
{
    private int _speed = 400;
    private float _angularSpeed = 12f;

    public override void _Process(double delta)
    {
        Rotation += Mathf.DegToRad(_angularSpeed) * (float)delta;
        Vector2 velocity = Vector2.Up.Rotated(Rotation) * _speed;

        Position += velocity * (float)delta;
    }
}
```

**Reply by AnAtlas** [reply]
I know this is an old comment, but replying for people who come after.

In the example position and rotation aren't capitalized. Could that be the issue?

**Reply by GrigorijeVigil** [reply]
For degrees, you need to use much larger values of angular_speed, eg 180 for Pi. Due to the slow turn rate, it seems that the icon just flies off the screen.

**Comment by abrossimow** [comment]
I highly recommend to use **typed variables** from the beginning. Static typing has two major advantages, especially for beginners:

1. typed variables will cause an error when assigned with wrong values. This will help you find bugs **a lot**.
2. typed variables will help the godot editor figure out **recommendations and symbol lookup** automatically!

Argument 2 to me is very important as it advocates for the godot editor being a modern IDE. Later, when your variables get more advanced and you use gdscript's included functions, you will be thankful for the quick tips the editor throws at you. So every typed variable will help you spend less time in the documentation. :)

Tbh: I don't know why this is not a standard in the documentation. Dynamic typing is all fun and easy, but imo the editor is such a great tool that offers a lot more "professional" help with static typing.
See here: https://docs.godotengine.org/en/stable/tutorials/scripting/gdscript/static_typing.html#doc-gdscript-static-typing

## tl/dr:
Learn to static type from the beginning. You will be glad you did in the future. So instead of doing this:
```
var speed = 400
var angular_speed = PI
```


do this:
```
var speed := 400
var angular_speed := PI

```
or better this:
```
var speed: int = 400
var angular_speed: float = PI

```

**Reply by Calinou** [reply]
> Tbh: I don't know why this is not a standard in the documentation. Dynamic typing is all fun and easy, but imo the editor is such a great tool that offers a lot more "professional" help with static typing.

See discussion in https://github.com/godotengine/godot-docs/issues/6295.

**Reply by AnAtlas** [reply]
Agreed, also speed and angular_speed should be const since the example doesn't mutate them.

**Comment by Jefferson-chidiox** [comment]
My image started running in circles immiediately after i ran `func _process(delta):
	rotation += angular_speed * delta` Turns out that the additional `	var velocity = Vector2.UP.rotated(rotation) * speed
	position += velocity * delta` Just makes it go up infinetly. I was using godot 4.4 at the time of this comment.

Wierdly, everything i run since the start of this tutorial stays by the side of the screen instead of being centered, regardless of where they are centered in my editor

**Reply by SriramAnanth2912** [reply]
I used the following code to center the sprite:
`var screen_size: Vector2 ## screen size of the viewport`

`func _ready() -> void:`
`	screen_size = get_viewport_rect().size`
`	global_position = screen_size / 2 `

**Reply by AnAtlas** [reply]
var velocity = Vector2.UP.rotated(rotation) * speed position += velocity * delta

How are you using velocity in the equation before you define it?

**Comment by TangyPrince** [comment]
I am very new to this, so please excuse me if I'm missing the obvious, but I'm having a hard time with the "position += velocity * delta" line. It's highlighted in red, and won't run. When I remove the line, the icon just rotates when I run it. I can't figure out what is going wrong, I've tried re-copying and pasting it and no luck...

**Reply by TangyPrince** [reply]
Ohhh. Figured it out. I needed to hit tab at the end of the line before it!

**Comment by SriramAnanth2912** [comment]
For anyone getting a Node configuration warning when dragging the Sprite2D, refer to this [#11267](https://github.com/godotengine/godot-docs/issues/11267). __This is not a critical warning.__

**Reply by leanne63** [reply]
Issue [#11267](https://github.com/godotengine/godot-docs/issues/11267) has been closed, without "fixing" the warning. Apparently, some ppl get this non-critical warning, while others don't.

TL;DR: if the warning bothers you, first create a Node2D as your root node. Then, add the Sprite2D node as a child of that one.

**Comment by tomas-chavarria** [comment]
From what I understand, the member variables are designed as "global" in the provided code. That is, they are not inside the function definition. 

This is generally considered bad practice in most coding backgrounds, but I was curious if there was a reason for this being encouraged here? Are the variables defined within a certain script confined to a given node so long as the script is attached?

**Reply by Calinou** [reply]
You need member variables to reference the same value in multiple functions, or to keep persistent state between two calls of the same function. It's not a bad practice per se, but you generally want to use local variables when a member variable isn't actually needed.

GDScript does not have "true" global variables like C or PHP do, but you can use [autoloads](https://docs.godotengine.org/en/stable/tutorials/scripting/singletons_autoload.html) to achieve a similar effect.

> Are the variables defined within a certain script confined to a given node so long as the script is attached?

If the same script is attached to multiple nodes, each node has its own value by default. However, if using a static variable (`static var x = "something"`), the value is shared across all nodes.

**Comment by oOWhattahWOo** [comment]
Question： Since `Position += velocity * delta` each frame, wouldn’t variable `delta` (dropped or fluctuating frames) make the icon drift off the intended circle?

**Reply by Calinou** [reply]
Multiplying the velocity by `delta` ensures the icon moves by a fixed distance regardless of framerate. If we did not multiply `velocity` by `delta`, the icon would move faster at higher framerates, or slower at lower framerates.

There is no risk of "drifting off" the circle here, as `delta` multiplies the velocity and can't change the direction by itself in this scenario.

**Reply by oOWhattahWOo** [reply]
Thank you for the reply, Calinou. I understand and agree with your explanation regarding the fixed distance (constant speed). However, my concern is actually about a different issue:
In the `_Process` function, we move the icon along a vector `Vector2.Up.Rotated(Rotation)`. If a frame takes a long time (high `delta`), the Rotation increases by a large amount instantly, and the object will travel a long straight line in that direction before it updates again.
Because of this coarse step (large rotation change + long linear move), won't the object permanently drift further away from its rotation center every time the framerate drops? It seems like the radius would keep increasing.

**Comment by Quirel** [comment]
So I decided to play a little with movement, starting postition etc. I've set initial position of sprite on my screen (not center of view port). 

Every time I run a program I've seen this sprite on the center of view port. Debugged this about 2 hours and then with help of google ai found we have **project splash screen**. So, it wasn't my sprite, just this splash screen uses exactly the same icon!

**Comment by 105913792** [comment]
I followed the first few steps but "Hello, World!" is not being printed in the Output Dock.

**Reply by 105913792** [reply]
The icon is also not rotating in place.

**Reply by 105913792** [reply]
I fixed it by clicking on the scroll/script button. It opened what had apparently been the attached script. To prevent confusion I detatched that script and attached the one I had wanted.

**Comment by night-kat** [comment]
It seems like the godot face is tracing a prolate trochoid instead of a circle, even when following the steps exactly and i cannot tell why. I code in rust but it did not work, so i tried GDScript and the example on this page, but it would still result in the same prolate trochoid.

**Comment by Xiarno** [comment]
A version of the tutorial should be written completely in C#.

The C# code is useful but with the explanation being in GDScript, it doesn't really help beginner like me.

**Comment by admin** [comment]
*Originally posted by [078111zz-oss](https://github.com/078111zz-oss):*

> ```gdscript
> extends Sprite2D
> 
> var speed = 400
> var angular_speed = PI
> 
> 
> func _process(delta):
> 		rotation += angular_speed * delta
> 		
> var velocity = Vector2.UP.rotated(rotation) * speed
> 	 position += velocity * delta
> ```
> 
> This last code shows an error in the last two lines.

**Reply by admin** [reply]
*Originally posted by [Calinou](https://github.com/Calinou):*

> > This last code shows an error in the last two lines.
> 
> The indentation is incorrect, it should be like this:
> 
> ```gdscript
> func _process(delta):
> 	rotation += angular_speed * delta
> 
> 	var velocity = Vector2.UP.rotated(rotation) * speed
> 	position += velocity * delta
> ```
> 
> Indentation matters in GDScript; you need to keep all lines within the function at the same indentation level (or greater, if nested within an `if` statement or similar).

**Reply by admin** [reply]
*Originally posted by [TisolaNox](https://github.com/TisolaNox):*

> Hello, 
> 
> in godot at the bottom of the script windows you should have a red line explaining the error  you can then click on the right side of it to show all errors 
> <img width="1865" height="243" alt="image" src="https://github.com/user-attachments/assets/51ee1023-4c55-4642-905c-b8cbe6f346c7" />
> 
> - Try to only use indentation, it a nice convention and it allow to have even spacing 
> - Both the variable declaration and the position value update should be made in the `_process`  function 
> so that the value can be updated each tick. By `in the function I mean to have one more indentationthan the function name. function is at 0  so try to set everything at one under it 
> -  You can technically add more than one indentation as long as it stays consistent 
> 
> ```
> extends Sprite2D
> 
> var speed = 400
> var angular_speed = PI
> 
> 
> func _process(delta):
> 	## MANY INDENTATION 
> 								rotation += angular_speed * delta
> 								
> 								var velocity = Vector2.UP.rotated(rotation) * speed
> 								position += velocity * delta
> 
> func _init():
> 	## ONLY ONE easier to read 
> 	print("HELLO WORLD")
> 	if speed == 400:
> 		## Two indentations because the print depends on the if statement (1 ind) that is in the init function (0 ind)
> 		print("current speed is 400")
> 
> 
> ```
> 
> cheers
